### DISCUSSION TITLE
How to compile a subtree with intermediate `opt` passes

### DISCUSSION THREAD

**Discussion by uduse** [SCons/scons#4198] [Q&A]
What I want to achieve:
The project is already buildable with scons. In the project, I have a file `math.cpp` that depends on `add.h` and some other files. There are other files in the project, such as `fruit.cpp` that I don't want to touch. The final executable depends on all these files. My goal is to apply additional LLVM passes such as `-sroa` to `math.cpp` and its dependencies. 

What I have tried:
I added `compilation_db` to SConsstruct, and built the project normally. This exported a compilation database as a JSON file. I used `python scons.py -Q --tree=prune math.o` to get a tree of dependencies for my math module. This gave me a list of `*.h` files. I replaced `*.h` with `*.o` and locate these symbols in the compilation database. For each command I located this way in the compilation database, I transformed it into three commands (1) original command but with `-emit-llvm` (2) `opt` with bunch of passes (3) output `*.o`. I manually tested these steps with one dependency and worked okay.

My questions are:
(1) When I run the commands one by one through Python using `subprocess.run`, the compilation is extremely slow. Also it seems it does not skip the files that do not need to be touched. I am not sure [using clang's Python interface](https://stackoverflow.com/questions/36652540/how-to-use-compile-commands-json-with-clang-python-bindings) could help. Does this sound like a good idea to you?
(2) Dependency resolution is already a part of scons, and I image it is possible to do this in scons. However, it is not obvious to me how to split one command action into three based on the dependency of a specific target. Can someone give me some hints how to do this?

**Comment by bdbaddog** [comment]
You'll have to explain more how you use -sroa .
I've not used it.

If you're running subprocess.run() chances are you're not using SCons properly.

Sounds like (and I could be entirely wrong so please correct me), you want to compile (at least) one source file with a different set of  compiler flags?

Maybe `-mllvm sroa` added to CFLAGS ?

**Reply by uduse** [reply]
I am not using `-sroa` specifically. I want this to work with arbitrary sequences of llvm passes. The goal is to replace the default optimization "-O3" with a user supplied (could be from scons.py's commandline arguments) llvm passes. [From what I've learned](https://groups.google.com/g/llvm-dev/c/Y1HHgpmBidM) this is only doable through explicit optimization through the usage of `opt`. So I have to go through the process of `*.cpp` ➡️ `*.ll` ➡️ `*.ll` (optimized through `opt`) ➡️ `*.o`, using `clang`, then `opt`, then `llc`.

Yes, I am aware that using `subprocess.run()` is stupid. I just don't know a better way yet.

> Sounds like (and I could be entirely wrong so please correct me), you want to compile (at least) one source file with a different set of compiler flags?

Yes, I want to compile a subset of source files with a different set of complier flags with different number of stages. Different set of compiler flags is doable with duplicating an environment then overwrite, but it is unclear to me how to split one such command into three.

**Reply by bdbaddog** [reply]
You don't have to duplicate an environment to have file specific values of env variables.

```env.SharedObject('math.cpp',CCFLAGS="$CCFLAGS -mllvm sroa"```

Would add `-mllvm sroa` (for example) only when compiling math.cpp

Can you provide a sample of the series of command lines you'd like to apply to math.cpp?

**Reply by uduse** [reply]
I don't think using `-mllvm` would work. The flags I'm using are a subset of [llvm optimization flags](https://llvm.org/docs/Passes.html), the clang frontend doesn't accept those. Say want to apply `-mem2reg -adce`. For each dependency of `math.cpp` and `math.cpp` itself, call it `x.cpp`, I will need to run `clang -emit-llvm -S -o x.ll x.cpp`, then `opt -S -o x.ll -mem2reg -adce x.ll`, then `llc -o x.o x.ll`.

**Reply by bdbaddog** [reply]
That'd be fairly easy.
You'll want to replace SHCXXCOM which currently has 1 Action. With 3. Should be able to do it with something like
```env.SharedObject('math.cpp',SHCXXCOM=['clang -emit-llvm -S -o ${TARGET.base}.ll $SOURCE', 'opt -S -o ${TARGET.base}.ll -mem2reg -adce ${TARGET.base}.ll', 'llc -o $TARGET ${TARGET.base}.ll'])```

Though you should probably not write to the same file you're reading.

So maybe:
```env.SharedObject('math.cpp',SHCXXCOM=['clang -emit-llvm -S -o ${TARGET.base}.ll $SOURCE', 'opt -S -o ${TARGET.base}-1.ll -mem2reg -adce ${TARGET.base}.ll', 'llc -o $TARGET ${TARGET.base}-1.ll'])```

**Reply by bdbaddog** [reply]
Also, you most likely would want to add a SideEffect() (if you didn't to create a new/additional emitter for that builder)
```
t = env.SharedObject('math.cpp',SHCXXCOM=['clang -emit-llvm -S -o ${TARGET.base}.ll $SOURCE', 'opt -S -o ${TARGET.base}-1.ll -mem2reg -adce ${TARGET.base}.ll', 'llc -o $TARGET ${TARGET.base}-1.ll'])

env.SideEffect(t, ['math.ll','math-1.ll'])
```

**Reply by uduse** [reply]
Thank you for your help! The LLVM passes the program will be using are dynamic. I think I will add `opt_passes = ARGUMENTS.get('OPT_PASSES')` and use python string concatenation to format  the command you just showed me. If I understand this correctly this won't trigger unnecessary rebuilds of other files. Additionally, I assume that this `env.SharedObject` call overrides the default behavior, right?

**Reply by bdbaddog** [reply]
Yes, it overrides the default actions for the SharedObject() builder, You may also want to set `SHCXXCOMSTR` to the same value so you'll see the command being run. (I forget if SHCXXCOMSTR="$SHCCCOM" or not.. if you see the non-default build commands when you build without overriding SHCXXCOMSTR then ignore this bit)

You will get rebuilds if you change the actions on files with the new actions. And any builders which depend on their targets should rebuild(recursively.. they're parents... etc)

**Reply by uduse** [reply]
Still doesn't work yet... More context, other files that I don't need to change are bunched with `math.cpp` to build a static library using `mylib = env.StaticLibrary('mylib', sources)`, where sources contains `math.cpp` and other files. 

I tried using 
```
env.SharedObject(
    "math.cpp",
    SHCXXCOM=[
        "clang -emit-llvm -S -o ${TARGET.base}.ll $SOURCE",
        "opt -S -o ${TARGET.base}-1.ll -reg2mem ${TARGET.base}.ll",
        "llc -o $TARGET ${TARGET.base}-1.ll",
    ],
)
```
but this gives me
```
Multiple ways to build the same target were specified for: workload/nbody_workload.os
```
Since other files are built with `env.StaticObject(source, ...)`, I tried to mimic this interface but couldn't figure out how to pass the right arguments... SCons either complains about multiple ways to build the same thing or building the same target in two environments. (I tried to find "the other environment", but couldn't find it either...)

**Reply by bdbaddog** [reply]
Yes remove math.cpp from the sources for that library and use the output from the SharedObject() builder as the source for that.
So adding to the previous example

```
math_object = env.SharedObject(
    "math.cpp",
    SHCXXCOM=[
        "clang -emit-llvm -S -o ${TARGET.base}.ll $SOURCE",
        "opt -S -o ${TARGET.base}-1.ll -reg2mem ${TARGET.base}.ll",
        "llc -o $TARGET ${TARGET.base}-1.ll",
    ],
)

sources = [s for s in sources if "math.cpp" not in s]
mylib = env.StaticLibrary('mylib', sources + math_object)
```

**Reply by uduse** [reply]
This almost works 😄  So first I tried it on some file `minus.cpp`, and it had missing header. If I understand it correctly, `SHCXXCOM` supposed to inherit those flags, but maybe it doesn't do so when there are three commands. I found `_CPPINCFLAGS` seems useful here, and it works for `minus.cpp`. I tried the same trick for `add.cpp` but this time though it found things to include, it reports `undefined reference` from a file that uses `add.cpp`. My guess is that my linking process is also missing some flags too. I tried adding linker related flags here and there without success.

What I have now:
```
add_obj = env.SharedObject(
    "add.cpp",
    SHCXXCOM=[
        "$CXX -emit-llvm -S -o ${TARGET.base}.ll $SOURCE $_CPPINCFLAGS",
        "opt -S -o ${TARGET.base}-2.ll -O0 ${TARGET.base}.ll",
        "llc -o $TARGET ${TARGET.base}-2.ll",
    ],
)
```

**Reply by bdbaddog** [reply]
Is it possible whatever optimizers you're running are dropping symbols from the generated .o?

**Reply by uduse** [reply]
It's unlikely, `-O0` should barely do anything. I think it has something to do with the fact that I'm doing cross-platform compilation. I need to get the flags in the first command right to give them access to everything... Are you aware of anything that might be different here if we cross-compile?

**Reply by bdbaddog** [reply]
There's lots different about cross compiling.
Do a regular compile of the object, run `nm` on it see if the reference is there, then `nm` the object you produce with your modified llvm command, see if the reference isn't there.

That's the best way to debug this issue.
Once you find what's causing the symbol to no longer be there, modify your more complicated compile flow to resolve.

This bit is out of scope of SCons.

**Comment by bdbaddog** [comment]
Really sounds like you just want to modify the `Action()`s which are used by the object builder for some source files?

**Comment by mwichmann** [comment]
If you're going to replace `SHCXXCOM` or some other core variable, it's useful to see what it was doing before and try to replicate.  These aren't "absolute", they might differ by compiler (that is, msvc usually does something different), but for starters, on my Linux system, in addition to `$SHCXXFLAGS $SHCCFLAGS`,  it contains  a "common" variable `$_CCCOMCOM` (also used by `SHCCCOM`, `CCCOM`, `CXXCOM`), which is defined as: `'$CPPFLAGS $_CPPDEFFLAGS $_CPPINCFLAGS'`.
The generated defines would seem to be fairly important...

I'd say pursuing changing the action will probably involve less hit-and-miss than trying to change the command variable.

**Reply by uduse** [reply]
This is actually the first thing we tried. We tried to add some code in `Action.py` but it really sounds like a bad idea (it's part of scons's source code). It maybe a better we do it in `SConstruct` passing some `executor`, but it's very unclear how I should proceed...

**Reply by bdbaddog** [reply]
Don't modify Action.py. That would be the wrong way to go about this.
